openrtb_ext/convert_up.go

```diff
@@ -0,0 +1,150 @@
+// object (req.ext.schain) to the OpenRTB 2.5 location (req.source.ext.schain). If the
+// OpenRTB 2.5 location is already present the OpenRTB 2.4 supply chain object is dropped.
+func moveSupplyChainFrom24To25(r *RequestWrapper) {
+	// read and clear 2.4 location
+	reqExt, _ := r.GetRequestExt()
+	schain24 := reqExt.GetSChain()
+	reqExt.SetSChain(nil)
+
+	// move to 2.5 location if not already present
+	sourceExt, _ := r.GetSourceExt()
+	if sourceExt.GetSChain() == nil {
+		sourceExt.SetSChain(schain24)
+	}
+}
+
+// moveSupplyChainFrom25To26 modifies the request to move the OpenRTB 2.5 supply chain
+// object (req.source.ext.schain) to the OpenRTB 2.6 location (req.source.schain). If the
+// OpenRTB 2.6 location is already present the OpenRTB 2.5 supply chain object is dropped.
+func moveSupplyChainFrom25To26(r *RequestWrapper) {
+	// read and clear 2.5 location
+	sourceExt, _ := r.GetSourceExt()
+	schain25 := sourceExt.GetSChain()
+	sourceExt.SetSChain(nil)
+
+	// move to 2.6 location if not already present
+	if schain25 != nil {
+		// source may be nil if moved indirectly from an OpenRTB 2.4 location, since the ext
+		// is not defined on the source object.
+		if r.Source == nil {
+			r.Source = &openrtb2.Source{}
+		}
+
+		if r.Source.SChain == nil {
+			r.Source.SChain = schain25
+		}
+	}
+}
+
+// moveGDPRFrom25To26 modifies the request to move the OpenRTB 2.5 GDPR signal field
+// (req.regs.ext.gdpr) to the OpenRTB 2.6 location (req.regs.gdpr). If the OpenRTB 2.6
+// location is already present the OpenRTB 2.5 GDPR signal is dropped.
+func moveGDPRFrom25To26(r *RequestWrapper) {
+	// read and clear 2.5 location
+	regsExt, _ := r.GetRegExt()
+	gdpr25 := regsExt.GetGDPR()
+	regsExt.SetGDPR(nil)
+
+	// move to 2.6 location
+	if gdpr25 != nil && r.Regs.GDPR == nil {
+		r.Regs.GDPR = gdpr25
+	}
+}
+
+// moveConsentFrom25To26 modifies the request to move the OpenRTB 2.5 GDPR consent field
+// (req.user.ext.consent) to the OpenRTB 2.6 location (req.user.consent). If the OpenRTB 2.6
+// location is already present the OpenRTB 2.5 GDPR consent is dropped.
+func moveConsentFrom25To26(r *RequestWrapper) {
+	// read and clear 2.5 location
+	userExt, _ := r.GetUserExt()
+	consent25 := userExt.GetConsent()
+	userExt.SetConsent(nil)
+
+	// move to 2.6 location
+	if consent25 != nil && r.User.Consent == "" {
+		r.User.Consent = *consent25
+	}
+}
+
+// moveUSPrivacyFrom25To26 modifies the request to move the OpenRTB 2.5 US Privacy (CCPA)
+// consent string (req.regs.ext.usprivacy) to the OpenRTB 2.6 location (req.regs.usprivacy).
+// If the OpenRTB 2.6 location is already present the OpenRTB 2.5 consent string is dropped.
+func moveUSPrivacyFrom25To26(r *RequestWrapper) {
+	// read and clear 2.5 location
+	regsExt, _ := r.GetRegExt()
+	usPrivacy25 := regsExt.GetUSPrivacy()
+	regsExt.SetUSPrivacy("")
+
+	// move to 2.6 location
+	if usPrivacy25 != "" && r.Regs.USPrivacy == "" {
```

Comment:
This is not necessarily the focus of this PR but I just noticed that you're doing a string compare on `usPrivacy25` here but for all other fields like `consent25` in the commit above it is a nil check. I took a look at the request wrapper and for every field we provide a `Get` and `Set` for, the field is a pointer. It seems like maybe this field should also be a pointer. We should investigate. Perhaps there's no impact but it is probably something worth looking into.

Reply:
Good point. I'll review the mxmcherry/openrtb field pointers. There's also reports that other fields should be pointers too, so I'll take a general review pass.